src/GraphQL/Telemetry/GraphQLTelemetryProvider.cs

```diff
@@ -44,11 +45,33 @@ public GraphQLTelemetryProvider(GraphQLTelemetryOptions options)
         _telemetryOptions = options;
     }
 
+    /// <summary>
+    /// Initializes a new instance that only executes when <see cref="OpenTelemetry.AutoInstrumentation.Initializer.Enabled"/>
+    /// is <see langword="true"/>, using the options from <see cref="OpenTelemetry.AutoInstrumentation.Initializer.Options"/>.
+    /// </summary>
+    private GraphQLTelemetryProvider()
+    {
+        _onlyWhenAutoTelemetryEnabled = true;
+        _telemetryOptions = OpenTelemetry.AutoInstrumentation.Initializer.Options ?? new GraphQLTelemetryOptions();
+    }
+
+    // make sure no accidental use of the default constructor by DI by making it private,
+    // accessible only through AutoTelemetryProvider, a shared static instance
+    internal static GraphQLTelemetryProvider? _autoTelemetryProvider;
+    internal static GraphQLTelemetryProvider AutoTelemetryProvider => _autoTelemetryProvider ??= new GraphQLTelemetryProvider();
+
     /// <inheritdoc/>
     public virtual float SortOrder => GraphQLBuilderExtensions.SORT_ORDER_CONFIGURATION;
 
     /// <inheritdoc/>
-    public virtual async Task<ExecutionResult> ExecuteAsync(ExecutionOptions options, ExecutionDelegate next)
+    public virtual Task<ExecutionResult> ExecuteAsync(ExecutionOptions options, ExecutionDelegate next)
+    {
+        return !_onlyWhenAutoTelemetryEnabled || OpenTelemetry.AutoInstrumentation.Initializer.Enabled
+            ? ExecuteInternalAsync(options, next)
+            : next(options);
```

Comment:
We can't change AddTelemetry to run before GraphQLBuilderBase adds the telemetry to the pipeline, but we can have the auto-added telemetry implementation check Enabled at runtime, so if it is subsequently disabled by a call to AddTelemetry, the auto-added telemetry is an allocation-free noop.

Reply:
Nope, RegisterDefaultServices occurs after the builder delegate.